### PR TITLE
enhancement(vrl): Add `geoip` enrichment table

### LINKED ISSUE #9283: Add geoip enrichment table
It would be of great help if VRL had support for a geoip function as this would make configs a lot lighter.

### Current Vector Version

```
vector 0.16.1 (x86_64-unknown-linux-gnu 739e878 2021-08-26)
```

### Use-cases
I am working with various different sources of logs, which tend to also have specific separate sinks and different formats. They often contain at least two fields where a part of the geoip information would be useful.

### Attempted Solutions
The way I currently do it is to have:
source -> transform[1..n] -> transform_geoip[1..n] -> transform (remove unused geoip fields) -> sink

Which translates to:
- source
- one or more transforms to parse and normalise the data.
- one or more geoip transforms, depending on how many fields there are
- final transform for that specific log source, which extracts the required geoip fields and removes the geoip object
- sink

Handling many different sources, this feels a bit cumbersome.

### Proposal

Having a VRL geoip function would make this a lot lighter, not requiring 1-n geoip transforms and no post-processing transform for removing unused fields and renaming fields.

### PR DESCRIPTION
**Pull request by ktff**
Closes  #9283.

Supports calls with condition like `{ "ip": .ip }`.

Also includes:
* Changes mentioned in https://github.com/vectordotdev/vector/pull/12803.
* `metro_code` is now  an integer. It was string before for consistency which I assume to mean was because all of the other fields were strings.
* `city_name` is now searched with configured `locale`. It was hard coded to search with "en" locale, which seems like a bug.

### CODE REVIEW

**Comment by ktff** [comment]
@jszwedko this is ready for review.

cc. @sim0nx

File: src/enrichment_tables/geoip.rs [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+        Ok(Box::new(Geoip::new(self.clone())?))
+    }
+}
+
+#[derive(Clone)]
+pub struct Geoip {
+    config: GeoipConfig,
+    dbreader: Arc<maxminddb::Reader<Vec<u8>>>,
+    last_modified: SystemTime,
+    indexes: Vec<(Case, Vec<String>)>,
```
Comment (StephenWakely) [inline src/enrichment_tables/geoip.rs:61]:
```suggestion

```

The indexes are not being used in the lookup, so we should be safe to just not store them at all.

File: src/enrichment_tables/geoip.rs [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+    /// # Errors
+    /// Errors if the fields are not in the table.
+    fn add_index(&mut self, case: Case, fields: &[&str]) -> Result<IndexHandle, String> {
+        match fields.len() {
+            0 => Err("IP field is required".to_string()),
+            1 => {
+                let index = IndexHandle(self.indexes.len());
+                self.indexes
+                    .push((case, fields.iter().map(|field| field.to_string()).collect()));
+                Ok(index)
```
Comment (StephenWakely) [inline src/enrichment_tables/geoip.rs:252]:
```suggestion
                Ok(0)
```

It's still useful to keep the error checks here, but we can remove the `push`.

File: website/cue/reference/remap/functions.cue [hunk 3]
```diff
@@ -77,6 +77,26 @@ remap: {
 		   match criteria. We recommend using date ranges as the _only_ criteria when the enrichment
 		   data set is very small.
 
+		For `geoip` enrichment tables this condition needs to be a VRL object with a single key-value pair
+		whose value needs to be a valid IP address. Example: `{"ip": .ip }`. If a return field is expected
```
Comment (StephenWakely) [inline website/cue/reference/remap/functions.cue:81]: Does it need to be ipv4 or does ipv6 work too?

File: website/cue/reference/remap/functions.cue [hunk 4]
```diff
@@ -77,6 +77,26 @@ remap: {
...
+		and without a value, `null` will be used. This table can return following fields:
+
+		* `autonomous_system_number`
+		* `autonomous_system_organization`
+		* `isp`
+		* `organization`
```
Comment (StephenWakely) [inline website/cue/reference/remap/functions.cue:87]: These fields are returned for an isp db and the other ones are for all other types of db. Can we make that clear here?

File: src/enrichment_tables/geoip.rs [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+    dbreader: Arc<maxminddb::Reader<Vec<u8>>>,
+    last_modified: SystemTime,
+    indexes: Vec<(Case, Vec<String>)>,
+}
+
+impl Geoip {
+    pub fn new(config: GeoipConfig) -> crate::Result<Self> {
+        Ok(Geoip {
+            last_modified: fs::metadata(&config.path)?.modified()?,
+            dbreader: Arc::new(maxminddb::Reader::open_readfile(config.path.clone())?),
```
Comment (StephenWakely) [inline src/enrichment_tables/geoip.rs:68]: Can we check here that the database is of a type that we support?

File: src/enrichment_tables/geoip.rs [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+    /// # Errors
+    /// Errors if no rows, or more than 1 row is found.
+    fn find_table_row<'a>(
+        &self,
+        case: Case,
+        condition: &'a [Condition<'a>],
+        select: Option<&[String]>,
+        index: Option<IndexHandle>,
+    ) -> Result<BTreeMap<String, Value>, String> {
+        self.find_table_rows(case, condition, select, index)?
```
[the remark this answers is not in the document]
Comment (ktff) [reply src/enrichment_tables/geoip.rs:208]: Well, current implementation will never find more than one row, but I see it making sense as future proofing.

File: src/enrichment_tables/geoip.rs (hunk 1 above)
Comment (ktff) [reply src/enrichment_tables/geoip.rs:61]: That may be the case at the moment, but if we are going to future proof `find_table_row` then it makes sense to also future proof this.

**Comment by ktff** [comment]
@StephenWakely I've added benchmarks, and it's seems Soak tests keep failing. The errors seems unrelated.

File: src/enrichment_tables/geoip.rs [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+                map.insert(key.to_string(), value.unwrap_or(Value::Null));
+            }
+        };
+
+        if self.has_isp_db() {
+            let data = self.dbreader.lookup::<Isp>(ip).ok()?;
+
+            add_field(
+                "autonomous_system_number",
+                data.autonomous_system_number.map(Into::into),
```
Comment (jszwedko) [inline src/enrichment_tables/geoip.rs:112]: Just a nit that all of these `add_field` calls seem to be calling `Into::into`. You could maybe DRY it up by making `add_field` take `Option<Into<Value>>` and doing it there.

File: src/enrichment_tables/geoip.rs [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+    fn lookup(&self, ip: IpAddr, select: Option<&[String]>) -> Option<BTreeMap<String, Value>> {
+        let mut map = BTreeMap::new();
+        let mut add_field = |key: &str, value: Option<Value>| {
+            if select
+                .map(|fields| fields.iter().any(|field| field == key))
+                .unwrap_or(true)
+            {
+                map.insert(key.to_string(), value.unwrap_or(Value::Null));
+            }
+        };
```
Comment (jszwedko) [inline src/enrichment_tables/geoip.rs:105]: Should we convert the `select` here to a `HashSet` to make the lookups cheaper?

**Review by jszwedko — COMMENTED** [review]
Nice work on this @ktff ! Thanks for adding the docs and making the adjustments mentioned in https://github.com/vectordotdev/vector/pull/12803.

**Review by jszwedko — APPROVED** [review]

File: src/enrichment_tables/geoip.rs (hunk 7 above)
Comment (ktff) [reply src/enrichment_tables/geoip.rs:112]: I would but closures can't be generic, and if we turn it into a function than two extra arguments would need to be passed which would negate gain from eliminating call to `Into::into`.

File: src/enrichment_tables/geoip.rs (hunk 8 above)
Comment (ktff) [reply src/enrichment_tables/geoip.rs:105]: For somewhat larger `select` lookups would be cheaper, but considering small size of `select` and the cost of creating `HashSet` just iterating over it should be slightly faster or comparable to using `HashSet`.

**Review by StephenWakely — APPROVED** [review]
This is looking nice!

**Comment by jszwedko** [comment]
Thanks for this @ktff !

**Pull request merged**